chain/chain/src/flat_storage_resharder.rs

```diff
@@ -980,11 +1066,193 @@ mod tests {
+            RawStateChangesWithTrieKey {
+                trie_key: account_mm_key,
+                changes: vec![RawStateChange { cause: StateChangeCause::InitialState, data: None }],
+            },
+            // Change: remove delayed receipt.
+            RawStateChangesWithTrieKey {
+                trie_key: delayed_receipt_1_key.clone(),
+                changes: vec![RawStateChange { cause: StateChangeCause::InitialState, data: None }],
+            },
+            // Change: remove buffered receipt.
+            RawStateChangesWithTrieKey {
+                trie_key: buffered_receipt_1_key.clone(),
+                changes: vec![RawStateChange { cause: StateChangeCause::InitialState, data: None }],
+            },
+        ];
+        manager
+            .save_flat_state_changes(block_hash, prev_hash, height, parent_shard, &state_changes)
+            .unwrap()
+            .commit()
+            .unwrap();
+
+        // Do resharding.
+        assert!(resharder.start_resharding(resharding_event_type, &new_shard_layout).is_ok());
+
+        // Validate integrity of children shards.
+        let flat_store = resharder.runtime.store().flat_store();
+        // Account 'oo' should exist only in the left child.
+        assert_eq!(
+            flat_store.get(left_child_shard, &account_oo_key.to_vec()),
+            Ok(account_oo_value.map(|val| FlatStateValue::inlined(&val)))
+        );
+        assert_eq!(flat_store.get(right_child_shard, &account_oo_key.to_vec()), Ok(None));
+        // Account 'vv' should exist with updated value only in the right child.
+        assert_eq!(flat_store.get(left_child_shard, &account_vv_key.to_vec()), Ok(None));
+        assert_eq!(
+            flat_store.get(right_child_shard, &account_vv_key.to_vec()),
+            Ok(account_vv_value.map(|val| FlatStateValue::inlined(&val)))
+        );
+        // Delayed receipt '1' shouldn't exist.
+        // Delayed receipt '0' should exist with updated value in both children.
+        for child in [left_child_shard, right_child_shard] {
+            assert_eq!(
+                flat_store.get(child, &delayed_receipt_0_key.to_vec()),
+                Ok(delayed_receipt_0_value_1.clone().map(|val| FlatStateValue::inlined(&val)))
+            );
+
+            assert_eq!(flat_store.get(child, &delayed_receipt_1_key.to_vec()), Ok(None));
+        }
+        // Buffered receipt '0' should exist with updated value only in the left child.
+        assert_eq!(
+            flat_store.get(left_child_shard, &buffered_receipt_0_key.to_vec()),
+            Ok(buffered_receipt_0_value_1.map(|val| FlatStateValue::inlined(&val)))
+        );
+        assert_eq!(flat_store.get(right_child_shard, &buffered_receipt_0_key.to_vec()), Ok(None));
+        // Buffered receipt '1' shouldn't exist.
+        for child in [left_child_shard, right_child_shard] {
+            assert_eq!(flat_store.get(child, &buffered_receipt_1_key.to_vec()), Ok(None));
+        }
```

Comment:
Ultimately we should check that the state root matches that on chain (which was calculated from memtrie resharding). I'm not sure how easy would it be to do in test though. We should definitely do it when replacing the temporary memtrie with new one though.